### PR TITLE
Add Perf_String_Substring_* benchmarks

### PR DESCRIPTION
**Pull request by nietras**
Draft benchmarks for Substring. cc @adamsitnik feel free to make any changes you like right away. For some reason I can no longer built this after updating latest main. In relation to https://github.com/dotnet/runtime/pull/62577

### CODE REVIEW

File: src/benchmarks/micro/libraries/System.Runtime/Perf.String.cs [hunk 1]
```diff
@@ -76,6 +76,7 @@ public string Remove_IntInt(string s, int i1, int i2)
         [Arguments("dzsdzsDDZSDZSDZSddsz", 0)]
         [Arguments("dzsdzsDDZSDZSDZSddsz", 7)]
         [Arguments("dzsdzsDDZSDZSDZSddsz", 10)]
+        [Arguments("dzsdzsDDZSDZSDZSddsz", 19)]
```
Comment (nietras) [inline src/benchmarks/micro/libraries/System.Runtime/Perf.String.cs:79]: probably will delete this change, so ignore it

File: src/benchmarks/micro/libraries/System.Runtime/Perf.String.cs [hunk 2]
```diff
@@ -255,7 +256,70 @@ private static char getStringCharNoInline(string str, int index)
...
+
+
+    //[BenchmarkCategory(Categories.Runtime, Categories.Libraries)]
+    //public class Perf_String_Substring_IntInt
+    //{
+    //    [Benchmark]
+    //    public void Substring_Full() => "abc".Substring(0, 3);
+
+    //    [Benchmark]
+    //    public void Substring_Empty() => "abc".Substring(0, 0);
```
Comment (nietras) [inline src/benchmarks/micro/libraries/System.Runtime/Perf.String.cs:320]: not sure these are of any interest but they showed that with the new changes the JIT can fully compile away substring code compared to before

**Comment by nietras** [comment]
build errors
```
C:\git\oss\performance\src\benchmarks\micro\Serializers\Binary_FromStream.cs(39,13): error SYSLIB0011: 'BinaryFormatter.Serialize(Stream, object)' is obsolete: 'BinaryFormatter serialization is obsolete and should not be used. See https://aka.ms/binaryformatter for more information.' [C:\git\oss\performance\src\benchmarks\micro\MicroBenchmarks.csproj]
C:\git\oss\performance\src\benchmarks\micro\Serializers\Binary_FromStream.cs(61,23): error SYSLIB0011: 'BinaryFormatter.Deserialize(Stream)' is obsolete: 'BinaryFormatter serialization is obsolete and should not be used. See https://aka.ms/binaryformatter for more information.' [C:\git\oss\performance\src\benchmarks\micro\MicroBenchmarks.csproj]
C:\git\oss\performance\src\benchmarks\micro\Serializers\Binary_ToStream.cs(40,13): error SYSLIB0011: 'BinaryFormatter.Serialize(Stream, object)' is obsolete: 'BinaryFormatter serialization is obsolete and should not be used. See https://aka.ms/binaryformatter for more information.' [C:\git\oss\performance\src\benchmarks\micro\MicroBenchmarks.csproj]
C:\git\oss\performance\src\benchmarks\micro\libraries\System.Runtime.Serialization.Formatters\Perf.BinaryFormatter.cs(30,13): error SYSLIB0011: 'BinaryFormatter.Serialize(Stream, object)' is obsolete: 'BinaryFormatter serialization is obsolete and should not be used. See https://aka.ms/binaryformatter for more information.' [C:\git\oss\performance\src\benchmarks\micro\MicroBenchmarks.csproj]
C:\git\oss\performance\src\benchmarks\micro\libraries\System.Runtime.Serialization.Formatters\Perf.BinaryFormatter.cs(38,32): error SYSLIB0011: 'BinaryFormatter.Deserialize(Stream)' is obsolete: 'BinaryFormatter serialization is obsolete and should not be used. See https://aka.ms/binaryformatter for more information.' [C:\git\oss\performance\src\benchmarks\micro\MicroBenchmarks.csproj]
```

**Comment by danmoseley** [comment]
Not sure why those build failures don't show up when I build locally. There should be something like this suppressing them though.
https://github.com/dotnet/runtime/blob/03e90a540cb2dfe5cab4086ea54ad5dd1f655749/src/libraries/System.Runtime/tests/System/ExceptionTests.cs#L238-L237

**Comment by nietras** [comment]
Suppressed the warning in the places this was missing. See PR. Then getting assets issue:
```
C:\Program Files\dotnet\sdk\7.0.100-alpha.1.21568.2\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(267,5): error NETSDK1005: Assets file 'C:\git\oss\performance\artifacts\obj\MicroBenchmarks\project.assets.json' doesn't have a target for 'net6.0'. Ensure that restore has run and that you have included 'net6.0' in the TargetFrameworks for your project. [C:\git\oss\performance\src\benchmarks\micro\MicroBenchmarks.csproj]
```
so switched to `netcoreapp5.0` in compilation. Tried rebuilding, cleaning, restoring didn't help. Didn't have these issues before updating to latest main.

**Comment by nietras** [comment]
Reported some results in https://github.com/dotnet/runtime/pull/62577#issuecomment-993493390

**Comment by danmoseley** [comment]
@nietras checking in on this one, do you believe it's ready to review (after resolving conflicts)?

**Comment by danmoseley** [comment]
@nietras I'm closing this one as it seems that string.Substring(int) and string.Substring(int, int) are already covered very similarly. Please let me know if that's not the case.

**Pull request closed without merging**